### PR TITLE
HPC-7904: 🛂 Setup authentication for resolvers

### PR DESCRIPTION
**Pull request by Pl217**
The main goal of this commit is to allow for usage of `@Permission` decorator, which allows our permission model to be used with resolver methods.

Also, there are multiple commits which deal with various improvements, not directly affecting auth, but I didn't want to open a lot of PRs, since the project is still in its inception and lots of changes are expected. Some of the notable changes:
- Update all dependencies to latest version, except `graphql`, which got released recently, but other dependencies like `type-graphql` [aren't ready for the upgrade](https://github.com/MichalLytek/type-graphql/issues/1100). Apollo server sees major version upgrade, going from `v2` to `v3`.
- `knex` version is pinned to the same version used in `hpc-api-core`, because there are type errors otherwise
- `hpc-api-core` dependency is added now that new model abstraction lib based on `knex` is added in https://github.com/UN-OCHA/hpc-api-core/pull/1 and some basic models defined in there. This PR will need https://github.com/UN-OCHA/hpc-api-core/pull/30 merged first, because some more models need to be defined, since I deleted models defined in this repo, because those defined in `hpc-api-core` will be used instead.
- Custom auth header parsing is introduced because plugins for [Basic](https://www.npmjs.com/package/@hapi/basic) and [Bearer](https://www.npmjs.com/package/hapi-auth-bearer-token) header parsing will not work together and I haven't found a plugin to support both
- Debugging is allowed with port forwarding enabled for debugging outside docker container. In order to attach debugger in chrome://inspect, you'll need to add localhost:9339 inside "Configure..." button popup dialog.
- Local copy of `hpc-api-core` repo is used, so new `yarn install-and-link dev` command is introduced that does the necessary linking.

### CODE REVIEW

File: src/data-providers/postgres/models/common/base.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export const createModel =
       getAll: () => table.select('*') as Promise<T[]>,
       getAllILike: <K extends keyof T>(k: K, v: T[K]) =>
         table.where(k as string, 'ilike', `%${v}%`) as Promise<T[]>,
-      create: (data: T) => table.insert(data),
+      create: (data: Partial<T>) => table.insert(data),
```
Comment (s0) [inline src/data-providers/postgres/models/common/base.ts]: If there are fields that are optional / not required in creation, we probably want to specify exactly which those are so that missing required fields are still provied...

That being said... probably not worthwhile fixing this as we'll be switching to the new abstraction library soon

... so feel free to ignore this

File: config/index.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-const config = {
+export const config = {
```
Comment (s0) [inline config/index.ts]: let's use the correct casing for global constants.

```suggestion
export const CONFIG = {
```

File: src/common-utils/log.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      console.error('Unrecognized logging mode:', config.logging.mode);
+    }
+  }
+  // TODO:
+  // once all jenkins scripts have been updated to use the LOG_MODE env var,
+  // this should be updated to use devServer instead of jenkinsScript,
+  // and when all local scripts have been updated, this can be removed.
+  if (config.name === 'dockerdev') {
+    return LOGGING_CONFIGS.devServer;
+  }
```
Comment (s0) [inline src/common-utils/log.ts]: this is not needed for the new API, as any use should give an apprioriate `config.name` that is valid in `isLoggingMode`

```suggestion
```

File: package.json [hunk 4]
```diff
@@ -16,9 +16,14 @@
   },
   "dependencies": {
     "apollo-server-hapi": "^2.25.0",
+    "bunyan": "^1.8.15",
     "class-validator": "^0.13.1",
+    "fp-ts": "^2.10.5",
     "graphql": "^15.5.0",
+    "io-ts": "^2.2.16",
```
Comment (s0) [inline package.json]: This will probably need to match what we're using in `hpc_service`, `hpc_cdm` and `hpc-api-core` to avoid type-checking errors etc...

File: src/common-libs/auth/permissions.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  | RequiredPermission
+  | 'anyone'
+  | 'noone';
+
+export type RequiredPermissionsConditionWithId =
+  | RequiredPermissionWithId
+  | 'anyone'
+  | 'noone';
+
+export const hasRequiredPermissions = (
```
Comment (s0) [inline src/common-libs/auth/permissions.ts]: looks like you dropped all the logic here that allows for the construction of a condition in Disjunctive Normal Form, is there a particular reason for not porting this across?

File: src/common-libs/auth/auth-checker.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { AuthChecker } from 'type-graphql';
+import { getLoggedInParticipant, getRoleGrantsForUser } from '.';
+import { createRootContextFromRequest } from '../../common-utils/logging';
+import Context from '../../domain-services/Context';
+
+export const authChecker: AuthChecker<Context> = async (
```
Comment (s0) [inline src/common-libs/auth/auth-checker.ts]: the majority of use cases will need to check more than just whether a user has a global role... in which case supplying role names is insufficient, as you also need additional context (such as the ID of object in question).

Beyond this, authorization checking should always be done with respect to permissions, not roles. Otherwise changing the permitted actions of any particular role means updating every endpoint rather than simply removing / adding a permission.

Given that you've mostly ported `calculatePermissionsFromRolesGrant`, is there any particular reason why you're not just using that and `hasRequiredPermissions`?

File: src/common-libs/auth/permission-decorator.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      }
+
+      if (
+        !(await auth.actionIsPermitted(
+          context.models,
+          context.request,
+          permissions
+        ))
+      ) {
+        throw new Error('No permission to perform this action');
```
Comment (s0) [inline src/common-libs/auth/permission-decorator.ts]: We probably want to start introducing a collection of errors for the different status codes we'll want to throw, so that this can be a 403 for example

File: src/common-libs/auth/permission-decorator.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+          throw new Error('Non global permissions require ID field provided');
+        }
+
+        permissions = {
+          id: args.id,
+          ...requiredPermissions,
+        };
+      } else {
+        permissions = requiredPermissions;
+      }
```
Comment (s0) [inline src/common-libs/auth/permission-decorator.ts]: I don't think I like this construction of permissions, for a number of reasons:

* It assumes that `args.id` will always be the ID of the target we want to authorize against
* We can only ever have 1 permission, with 1 ID (what if we wanted to authenticate against a governingEntity, and needed to allow either a grant for the governingEntity, or the plan that it belongs to).
* We could have better type-checking guarantees.

As an alternatve, I suggest that you make `requiredPermissions` a function that takes `ResolverData<Context>` as an argument, and returns a `Promise<RequiredPermissionsCondition>`, which you then call inside the body of the decorator. That allows the ID etc... to be precisely determined (in a type-checked manner), and also allows for the re-introduction of more advanced permission conditions,

File: docs/AUTH.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+The auth logic is mostly ported from v3 API ([hpc_service](https://github.com/UN-OCHA/hpc_service)) and available for use with two decorators:
+
+## `Authorized()`
+
+This decorator is [natively supported](https://typegraphql.com/docs/authorization.html) by TypeGraphQL and enables to restrict access to logged-in users only.
+
+When used with parameters, this decorator can be used to require users to have one of the listed roles. For example, decorating with `@Authorized('projectsAdmin')` allows only projects admin to access resource.
+
+**Note: Only global roles can be used with this decorator.**
```
Comment (s0) [inline docs/AUTH.md]: I'm not sure that we have a concrete need for this decorator. If something requires a authorized acces, then it should require a specific permission, if not then it should be open to anyone to access (subject to rate-limiting of course, but that's a different matter).

I also see a potential risk that developers may incorrectly use `Authorized` instead of `Permission` in a number of cases (ie. new developers that have not yet fully familiarized themselves with the system).

**Review by s0 — CHANGES_REQUESTED** [review]
Few comments here.

There is of course quite a lot of duplication here from the old service (which does make it difficult to review exactly how it differs), but I've given that code in particular at least a quick once-over.

I'm wondering now if it makes sense though to move all that code in `hpc_service` that we want to share into `hpc-api-core` before merging this PR, and basing this stuff on-top of that, avoiding the duplication in the first place. There will certainly be a bunch of changes that need to be made to make it compatible for both use-cases, but that could then be done in separate PRs that make it clear how this code is changing.

File: config/index.ts (hunk 2 above)
Comment (Pl217) [reply config/index.ts]: Done!

File: src/common-utils/log.ts (hunk 3 above)
Comment (Pl217) [reply src/common-utils/log.ts]: This code is moved to API core in https://github.com/UN-OCHA/hpc-api-core/pull/11

File: src/common-libs/auth/permissions.ts (hunk 5 above)
Comment (Pl217) [reply src/common-libs/auth/permissions.ts]: This code is moved to API core in https://github.com/UN-OCHA/hpc-api-core/pull/11 and conditions in Disjunctive Normal Form can now be constructed.

File: src/common-libs/auth/auth-checker.ts (hunk 6 above)
Comment (Pl217) [reply src/common-libs/auth/auth-checker.ts]: This decorator is now removed and only `@Permission` decorator is left.

File: src/common-libs/auth/permission-decorator.ts (hunk 7 above)
Comment (Pl217) [reply src/common-libs/auth/permission-decorator.ts]: I wanted to "port" `errors` module from `hpc_service`, but it is heavily tied to `resitfy`, so I didn't do it in this PR. I agree we should do this soon.

File: src/common-libs/auth/permission-decorator.ts (hunk 8 above)
Comment (Pl217) [reply src/common-libs/auth/permission-decorator.ts]: I accepted your suggestion to create a function that takes `ResolverData<Context>` as an argument, and returns a `Promise<RequiredPermissionsCondition>`.

File: docs/AUTH.md (hunk 9 above)
Comment (Pl217) [reply docs/AUTH.md]: Removed `@Authorized()` decorator altogether.

**Comment by Pl217** [comment]
> I'm wondering now if it makes sense though to move all that code in `hpc_service` that we want to share into `hpc-api-core` before merging this PR, and basing this stuff on-top of that, avoiding the duplication in the first place. There will certainly be a bunch of changes that need to be made to make it compatible for both use-cases, but that could then be done in separate PRs that make it clear how this code is changing.

I moved the common auth code to https://github.com/UN-OCHA/hpc-api-core/pull/21

File: src/common-libs/auth/permission-decorator.ts (hunk 7 above)
Comment (Pl217) [reply src/common-libs/auth/permission-decorator.ts]: This is done now

File: src/domain-services/location/location-service.ts [hunk 10]
```diff
@@ -1,23 +1,33 @@
...
 @Service()
 export class LocationService {
   async findById(
-    models: DbModels,
+    models: Database,
     id: number
-  ): Promise<LocationTable | undefined> {
-    const location = await models.location.getOne(id);
+  ): Promise<InstanceDataOfModel<Database['location']>> {
+    const location = await models.location.findOne({
```
Comment (s0) [inline src/domain-services/location/location-service.ts]: `.get()` could be used here instead

File: src/domain-services/plans/plan-service.ts [hunk 11]
```diff
@@ -1,42 +1,51 @@
...
 export class PlanService {
-  async findById(models: DbModels, id: number) {
-    const plan = await models.plans.plan.getOne(id);
-
-    if (plan[0] === undefined) {
+  async findById(
+    models: Database,
+    id: number
+  ): Promise<{ id: PlanId; name?: string | null }> {
+    const plan = await models.plan.findOne({
```
Comment (s0) [inline src/domain-services/plans/plan-service.ts]: `get()` could be used instead

File: src/domain-services/plans/plan-service.ts [hunk 12]
```diff
@@ -1,42 +1,51 @@
...
-    const latestPlanVersion = (await models.plans.planVersion.table().where({
-      planId: planId,
-      latestVersion: true,
-    })) as any as PlanVersionTable[];
+    const latestPlanVersion = await models.planVersion.findOne({
+      where: {
+        planId,
+        latestVersion: true,
+      },
+    });
```
Comment (s0) [inline src/domain-services/plans/plan-service.ts]: If this endpoint is unauthenticated, we instead want to get the `current` plan, to prevent inadvertently allowing access to unpublished plan data. Beyond that, if there is no `currentPlanVersion` that matches the requirements, we should 404 instead.

File: config/index.ts [hunk 13]
```diff
@@ -21,6 +21,7 @@ export const config = {
   uploadTo: 'aws',
   name: process.env.NODE_ENV,
   authBaseUrl: process.env.AUTHBASE_URL,
+  basicAuthUser: process.env.BASIC_AUTH_USER,
```
Comment (s0) [inline config/index.ts]: we probably want to delete that as we've hard-coded the expected auth user for HID in `hpc-api-core` if i remember correctly.

If not, that's something we want to do, and also this change doesn't belong in this commit ;)

File: src/common-libs/auth/index.ts [hunk 14]
```diff
@@ -1,14 +1,69 @@
...
+    throw new Error('Bad HTTP authentication header format');
+  }
+
+  let credentials: string | null = null;
+  const schemePart = pieces[0];
+  const credentialsPart = pieces[1];
+
+  switch (schemePart.toLowerCase()) {
+    case 'basic':
+      credentials = parseBasic(credentialsPart).password;
```
Comment (s0) [inline src/common-libs/auth/index.ts]: here we want to check that the username is what we expect it to be (i.e. `BASIC_AUTH_USER`, specified in `hpc-api-core`). If the username differs to this, we should throw an error that says something along the lines of `Client Authentication is not supported in the v4 API`. (which is the traditional use of Basic auth for our old endpoints).

We could choose to support them later, but to be honest i'd prefer to introduce a more robust system for authentication of bots / peer services and get people to migrate to that if they want to use the v4 endpoints.

File: src/common-libs/auth/index.ts [hunk 15]
```diff
@@ -1,14 +1,69 @@
...
+    return null;
+  }
+
+  const pieces = req.headers.authorization.split(/\s+/);
+
+  if (!pieces || pieces.length !== 2) {
+    throw new Error('Bad HTTP authentication header format');
+  }
+
+  let credentials: string | null = null;
```
Comment (s0) [inline src/common-libs/auth/index.ts]: rather then setting a veriable and returning it at the end of the function, i'd prefer to just return directly inside the switch/case, and get rid of the breaks and trailing return statement.

File: src/common-libs/auth/permission-decorator.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+      const { context } = resolverData;
+
+      if (!(await actionIsPermitted(permissions, context))) {
+        throw new ForbiddenError('No permission to perform this action');
+      }
+
+      return next();
+    }
+  );
+}
```
Comment (s0) [inline src/common-libs/auth/permission-decorator.ts]: The commit message for 83852014da82ce19462d617698dc21eca394a678 seems to misalign with the actual implementation of `@Permission`, which instead seems to take a function and return the required permissions, which I like much more than what the commit describes (and I think is what I suggested when Joseph originally implemented this).

I would perhaps extend this to also support simply taking a `RequiredPermissionsCondition` directly too, and checking whether the given `requiredPermissions` is a function, and treating appropriately, which would reduce the code needed when using this decorator in simple permissions such as global permissions only.

**Review by s0 — CHANGES_REQUESTED** [review]
Absolutely love it! Great work!!!

Few minor comments, but nothing serious. Very excited to get this merged!

File: src/common-libs/auth/permission-decorator.ts (hunk 16 above)
Comment (Pl217) [reply src/common-libs/auth/permission-decorator.ts]: I forgot to update the commit message after switching to function inputs. I am not very fond of allowing multiple ways of specifying the same thing, but I will respect the reviewer :) I can see me nitpicking later when giving reviews: "Here you only use `global` permissions, don't wrap inside the function" :smile:

File: src/common-libs/auth/index.ts (hunk 15 above)
Comment (Pl217) [reply src/common-libs/auth/index.ts]: Done. This was used because I was forced :smile: I was not realizing that there is another way around ESLint's `no-case-declarations`. But, you can just wrap the block in curly braces to fix this error.

File: src/common-libs/auth/index.ts (hunk 14 above)
Comment (Pl217) [reply src/common-libs/auth/index.ts]: Done

File: config/index.ts (hunk 13 above)
Comment (Pl217) [reply config/index.ts]: You are right on both points :)

File: src/domain-services/plans/plan-service.ts (hunk 12 above)
Comment (Pl217) [reply src/domain-services/plans/plan-service.ts]: It was like that when i got here :smile: Thanks for catching this. We now need https://github.com/UN-OCHA/hpc-api-core/pull/49 for 404 errors

File: src/common-libs/auth/permission-decorator.ts (hunk 16 above)
Comment (s0) [reply src/common-libs/auth/permission-decorator.ts]: > "Here you only use global permissions, don't wrap inside the function" 😄

Sounds like a very reasonable review comment haha ;) Could maybe even make a linting rule for it! :-D

File: src/domain-services/location/location-service.ts (hunk 10 above)
Comment (Pl217) [reply src/domain-services/location/location-service.ts]: Wonder where I saw this review comment :smile:

Comment (s0) [reply src/domain-services/location/location-service.ts]: 😉

File: src/common-libs/auth/permission-decorator.ts (hunk 16 above)
Comment (Pl217) [reply src/common-libs/auth/permission-decorator.ts]: Right. Why keep simple interface, when we can make it possible to use it multiple ways, and then restrict in some cases :D

File: src/common-libs/auth/index.ts [hunk 17]
```diff
@@ -1,14 +1,72 @@
...
+
+  const schemePart = pieces[0];
+  const credentialsPart = pieces[1];
+
+  switch (schemePart.toLowerCase()) {
+    case 'basic': {
+      const credentials = parseBasic(credentialsPart);
+
+      if (credentials.username !== BASIC_AUTH_USER) {
+        throw new Error('Client Authentication is not supported in the v4 API');
```
Comment (s0) [inline src/common-libs/auth/index.ts]: This will produce a 500 error right? Let's use a BadRequest (400) error for this instead, as it is a misuse of the API that we should expect to need to handle.

**Review by s0 — COMMENTED** [review]
Looking good. One minor comment to address, which you can do at the point at which you update this to include the most recent `hpc-api-core`

**Review by s0 — APPROVED** [review]
